### DISCUSSION TITLE
"Smart" route URLs caching

### DISCUSSION THREAD

**Discussion by Voltra** [Z3d0X/filament-fabricator#116] [Ideas]
As of now, the route URLs caching system clears everything on deletion or update, and caches everything single route at one.

When a website grows, the number of pages can become quite large and this can slow down things quite a lot when editing pages.

It'd be better to have a tree-like caching structure to reflect the tree-like structure of the URLs (given the parent/child relationships in place). This will be a complete structural rewrite for the URLs system.

I suggest that a change in a parent page invalidates its own cache as well as the caches for the child routes.

I also suggest 5 new extension points:
* `Page#getUrlCacheKey($args = [])` To get the cache key that depends on developer-provided arguments (merged with default args)
* `Page#getUrl($args = [])` To get the URL that depends on developer-provided arguments (merged with default args, cached by the associated cache key)
* `Page#getDefaultUrlArgs()` To get the default arguments to merge with for the extension points above
* `Page#getAllUrlCacheKeys()` To get all the generatable URL cache keys for this specific page (e.g. translatable pages) OR `Page#getAllCacheKeysArgs()` to get all the argument sets for the generatable URLs and their cache keys
* `Page#getAllUrls()` To get all the generatable URLs for this specific page (e.g. translatable pages)

I also suggest maintaining a URL to model-id cache (that can be partially updated when a model is deleted or saved) that would make route argument lookups much faster.

**Comment by Z3d0X** [comment]
Interesting Idea. I will dedicate some time to thoroughly consider this when i get some time.

**Comment by Ray-vV** [comment]
I'm experiencing similar issues with this package. I'm developing a website which will have around 157,000 fabricator pages. 

When editing and saving one of these pages it takes around 12 seconds for the save to complete, which is way too long for pages that will be edited frequently.

I noticed that upon saving, a query is executed, which fetches all pages where parent_id === null?

Since I'm currently not using the parent_id  at all, it's fetching 157,000 pages where parent_id === null. Why is this? Why not just update the page that is actually being updated?
EDIT2: EXAMPLES:
Here are some examples to illustrate the issues I'm experiencing.

This is in a production environment. Here, it even takes longer (37 seconds)

![Schermafbeelding 2023-07-04 112124](https://github.com/Z3d0X/filament-fabricator/assets/17530532/6d1d2eeb-e9b3-42b0-97f2-02c660dda0ca)
Here you can see that all IDs for pages are being fetched in the '_select * from pages where pages.parent_id in_' query.
This is the reason the save takes such a long time, along with the '_select id, slug, title from pages where parent_id is null_' query.
(I only took a pic of the start and the end of the list of IDs, way too long to add here)
![Schermafbeelding 2023-12-01 115309](https://github.com/Z3d0X/filament-fabricator/assets/17530532/c44d75e8-5642-4b44-a624-00b7d87e175d)
![Schermafbeelding 2023-12-01 115357](https://github.com/Z3d0X/filament-fabricator/assets/17530532/f529f15f-e5b0-422a-8953-ae4e78bb3991)


Been trying to find a workaround for this all day, but haven't found anything yet. Is there a way to turn this off? 

UPDATE: 
I have tried adding parent_id to fabricator pages where applicable. Hoped it might speed up the process slightly.
Somehow, this cause all edit fabricator pages to load indefinitely (and ultimately crash, hitting the memory limit). Same when attempting to visit the corresponding page.

**Reply by Voltra** [reply]
That is indeed the exact problem this proposal aims to fix

**Comment by Voltra** [comment]
See #119

**Reply by Ray-vV** [reply]
Awesome work! Hopefully it can be integrated into the package soon. I could help with testing if you'd like.
